after/replace_forbidden_char.py

```diff
@@ -0,0 +1,82 @@
+#!/usr/bin/env python
+import os
+import sys
+
+dbname = sys.argv[1]
+config_file = sys.argv[2]
+from trytond.config import config as CONFIG
+CONFIG.update_etc(config_file)
+
+from trytond.pool import Pool
+from trytond.transaction import Transaction
+import logging
+
+Pool.start()
+pool = Pool(dbname)
+pool.init()
+
+context = {}
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.DEBUG)
+ch = logging.StreamHandler(sys.stdout)
+ch.setLevel(logging.DEBUG)
+formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
+ch.setFormatter(formatter)
+logger.addHandler(ch)
+
+from sql.functions import Function
+from trytond.model import fields
+# from psycopg2.errors import UndefinedColumn
+
+class StringAgg(Function):
+    __slots__ = ()
+    _function = 'STRING_AGG'
+
+
+with Transaction().start(dbname, 0, context=context):
+    Model = pool.get('ir.model')
+    ModelField = pool.get('ir.model.field')
+
+    model_h = ModelField.__table__()
+    model_field_h = ModelField.__table__()
+
+    cursor = Transaction().connection.cursor()
+
+
+    query = 'SELECT * FROM pg_catalog.pg_tables'
+    cursor.execute(query)
+    tables = [name for _, name, _, _, _, _, _, _ in cursor.fetchall()]
+
+    sql_where = (model_field_h.ttype == 'char')
+    query = model_field_h.select(model_field_h.model, StringAgg(model_field_h.name, ','), group_by=model_field_h.model, where=sql_where)
+    cursor.execute(*query)
+
+    for model_id, model_fields in cursor.fetchall():
```

Comment:
Trobo que et compliques molt la vida....

Per què no fas:

for field in ModelField.search([('ttype', '=', 'char')]):
    ....

Ja entenc que intentes fer tots els updates a la vegada però tal com ho estàs fent estàs fent els updates de tots els camps sí o sí (cosa que serà molt lenta en les bases de dades més grans), però en realitat només hauries de fer update si el camp conté algun dels caràcters especials que no hauria de tenir...

Potser ja ho has provat i has vist que el que has fet és més ràpid.... no sé.

Reply:
No et negaré que a mesura que anava fent l'script hem trobava casos a controlar-ho...

La proposta de search que fas, l'estava implementant i l'he avortat. => vull en una query saber el "model + tots els camps". Si no hauré de fer el group o fer la cerca a partir de model i obtenir els camps.... i tb hem complico la vida ;)

L'observació que fas de rendiment, m'has fet pensar-hi, i NO cal fer un update de tots els registres d'una taula, quan en aquella taula son X elements (per exemple, 10mil productes i només son 100 els productes a fer el update).

Actualitzo l'script amb aquest punt amb WHERE ids i un fix que he vist.

Es podria millorar i per exemple si a una taula tenim els camps char "name" i "code", i només té que actualitzar el camp "name", no faci el "code". Però aqui crec que ja ens compliquem el codi....

Segurament ens trobarem només en taules que s'han importat dades, com tercers o productes. En el cas que m'he trobat era en linies d'extracte (potser per la importació).

El que podem trobar-nos i segurament caldrà anar depurant, es per exemple a esale, importer, nexus o similars, .... que ens vinguin dades amb aquests caràcters i ens salti ForbiddenCharValidationError. (caldrà contolar-ho mòdul per mòdul)

Reply:
El WHERE se't tornarà molt complicat gran si vols fer-ho per tots els registres a la vegada. Continuo pensant que és més senzill fer camp a camp. El codi és més curt, raonablement eficient i més senzill. IMO.